starter-code/routes/auth.js

```diff
@@ -0,0 +1,116 @@
+const express = require("express");
+const router = express.Router();
+
+router.get("/signup", (req, res, next) => {
+  res.render("auth/signup");
+});
+
+router.get("/login", (req, res, next) => {
+  res.render("auth/login");
+});
+
+router.get("/secret", (req, res, next) => {
+  res.render("secret");
+});
+
+// User model
+const User = require("../models/user");
+
+// BCrypt to encrypt passwords
+const bcrypt = require("bcrypt");
+const bcryptSalt = 10;
+
+router.post("/signup", (req, res, next) => {
+
+
+  const username = req.body.username;
+  const password = req.body.password;
+  const salt = bcrypt.genSaltSync(bcryptSalt);
+  const hashPass = bcrypt.hashSync(password, salt);
+
+  if (username === "" || password === "") {
+    res.render("auth/signup", {
+      errorMessage: "Indicate a username and a password to sign up"
+    });
+    return;
+  }
+
+  User.findOne({
+      "username": username
+    })
+    .then(user => {
+      if (user !== null) {
+        res.render("auth/signup", {
+          errorMessage: "The username is already taken!"
+        });
+        return;
+      }
+
+      const salt = bcrypt.genSaltSync(bcryptSalt);
+      const hashPass = bcrypt.hashSync(password, salt);
+
+      User.create({
+          username,
+          password: hashPass
+        })
+        .then(() => {
+          res.redirect("/");
+        })
+        .catch(error => {
+          console.log(error);
+        })
+    })
+    .catch(error => {
+      next(error);
+    })
+});
+
+router.post("/login", (req, res, next) => {
+  const theUsername = req.body.username;
+  const thePassword = req.body.password;
+
+  if (theUsername === "" || thePassword === "") {
+    res.render("auth/login", {
+      errorMessage: "Please enter both, username and password to sign up."
+    });
+    return;
+  }
+
+  User.findOne({
+      "username": theUsername
+    })
+    .then(user => {
+      if (!user) {
+        res.render("auth/login", {
+          errorMessage: "The username doesn't exist."
+        });
+        return;
+      }
+      if (bcrypt.compareSync(thePassword, user.password)) {
+        // Save the login in the session!
+        req.session.currentUser = user;
+        res.redirect("/");
+      } else {
+        res.render("auth/login", {
+          errorMessage: "Incorrect password"
+        });
+      }
+    })
+    .catch(error => {
+      next(error);
+    })
+});
+
+router.use((req, res, next) => {
+  if (req.session.currentUser) {
+    next();
+  } else {
+    // res.redirect("/login"); OR REDIRECT TO THE 'SECRET' PAGE
+    res.redirect("/secret");
+  }
+});
+router.get("/secret", (req, res, next) => {
+  res.render("secret");
+});
+
```

Comment:
For the third iteration, you can write a middleware function and then pass it to the routes that need to be protected like for instance `const auth = (req, res, next) => {
  if (!req.session.user) {
    res.redirect("/auth/login");
    return;
  }
  next();
};

router.get("/private", auth, (req, res, next) => {
  res.render("auth/private");
});

router.get("/main", auth, (req, res, next) => {
  res.render("auth/main");
});`